### PR TITLE
incr.comp.: Move task result fingerprinting into DepGraph.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR 
- makes the DepGraph store all `Fingerprints` of task results,
- allows `DepNode` to be marked as input nodes, 
- makes HIR node hashing use the regular fingerprinting infrastructure,
- removes the now unused `IncrementalHashesMap`, and
- makes sure that `traits_in_scope_map` fingerprints are stable.

r? @nikomatsakis 
cc @alexcrichton

### CODE REVIEW

File: src/librustc_incremental/persist/preds/mod.rs [hunk 1]
```diff
@@ -66,7 +66,7 @@ impl<'q> Predecessors<'q> {
         // Reduce the graph to the most important nodes.
         let compress::Reduction { graph, input_nodes } =
             compress::reduce_graph(&query.graph,
-                                   |n| HashContext::is_hashable(tcx, n),
+                                   |n| n.kind.is_input(),
```
Comment (nikomatsakis) [inline src/librustc_incremental/persist/preds/mod.rs]: neat

File: src/librustc/dep_graph/graph.rs [hunk 2]
```diff
@@ -26,7 +26,8 @@ use super::edges::{DepGraphEdges, DepNodeIndex};
 
 #[derive(Clone)]
 pub struct DepGraph {
-    data: Option<Rc<DepGraphData>>
+    data: Option<Rc<DepGraphData>>,
+    fingerprints: Rc<RefCell<FxHashMap<DepNode, Fingerprint>>>
```
Comment (nikomatsakis) [inline src/librustc/dep_graph/graph.rs]: can we use a `DepNodeIndex` and make this a vector?

Comment (nikomatsakis) [reply src/librustc/dep_graph/graph.rs]: I guess not, because it is still used when incremental is disabled -- can you add some comments to that effect?

**Review by nikomatsakis — APPROVED** [review]
r=me modulo nits

File: src/librustc/dep_graph/graph.rs (hunk 2 above)
Comment (michaelwoerister) [reply src/librustc/dep_graph/graph.rs]: Maybe we can at some point, but I didn't want to complicate things. Note that not all `DepNode` have an associated `Fingerprint` (e.g. anonymous nodes). Last time I checked, there where 15% anonymous nodes, which would result in an OK fill rate for a vector.

Anyway, I want to revisit this once we see how the rest of the system shakes out.

Comment (michaelwoerister) [reply src/librustc/dep_graph/graph.rs]: Plus I'll add a comment for the meantime `:)`

**Comment by michaelwoerister** [comment]
@bors r=nikomatsakis

Comment added.

**Pull request merged**